core/src/main/scala/kafka/server/ClientQuotaManager.scala

```diff
@@ -54,11 +58,73 @@ object ClientQuotaManagerConfig {
   val DefaultQuotaWindowSizeSeconds = 1
   // Purge sensors after 1 hour of inactivity
   val InactiveSensorExpirationTimeSeconds  = 3600
+
+  val UnlimitedQuota = Quota.upperBound(Long.MaxValue)
+  val DefaultClientIdQuotaId = QuotaId(None, Some(ConfigEntityName.Default))
+  val DefaultUserQuotaId = QuotaId(Some(ConfigEntityName.Default), None)
+  val DefaultUserClientIdQuotaId = QuotaId(Some(ConfigEntityName.Default), Some(ConfigEntityName.Default))
+}
+
+object QuotaTypes {
+  val NoQuotas = 0
+  val ClientIdQuotaEnabled = 1
+  val UserQuotaEnabled = 2
+  val UserClientIdQuotaEnabled = 4
 }
 
+object QuotaId {
+
+  /**
+   * Sanitizes user principal to a safe value for use in MetricName
+   * and as Zookeeper node name
+   */
+  def sanitize(user: String): String = {
+    val encoded = URLEncoder.encode(user, StandardCharsets.UTF_8.name)
+    val builder = new StringBuilder
+    for (i <- 0 until encoded.length) {
+      encoded.charAt(i) match {
+        case '*' => builder.append("%2A") // Metric ObjectName treats * as pattern
+        case '+' => builder.append("%20") // Space URL-encoded as +, replace with percent encoding
```

Comment:
URLEncoder doesn't encode '*'. It does encode '+', but space is then encoded as '+'. So we replace '+' with %20 which is percent encoding for space.